File: aegis/frontend/src/test/utils/pagination.test.ts
import {
  defaultPaginationParams,
  validatePaginationParams,
  paginateArray,
  calculateOffset,
  calculateTotalPages,
  getPageInfo,
  createPaginationMeta,
  encodeCursor,
  decodeCursor,
  validatePageSize,
  getDefaultPageSize,
  getMaxPageSize,
  isValidPage,
  isValidPageSize,
  generatePageNumbers,
  getPageSizeOptions,
  needsPagination,
  getPaginationSummary,
} from '../../utils/pagination';

describe('Pagination Utils', () => {
  describe('defaultPaginationParams', () => {
    it('should return default pagination parameters', () => {
      const params = defaultPaginationParams();
      expect(params.page).toBe(1);
      expect(params.page_size).toBe(10);
    });
  });

  describe('validatePaginationParams', () => {
    it('should validate correct parameters', () => {
      const params = { page: 2, page_size: 20 };
      const result = validatePaginationParams(params);
      expect(result.page).toBe(2);
      expect(result.page_size).toBe(20);
    });

    it('should fix negative page', () => {
      const params = { page: -1, page_size: 20 };
      const result = validatePaginationParams(params);
      expect(result.page).toBe(1);
    });

    it('should fix zero page', () => {
      const params = { page: 0, page_size: 20 };
      const result = validatePaginationParams(params);
      expect(result.page).toBe(1);
    });

    it('should fix negative page size', () => {
      const params = { page: 1, page_size: -5 };
      const result = validatePaginationParams(params);
      expect(result.page).toBe(1);
      expect(result.page_size).toBe(1);
    });

    it('should cap page size at 100', () => {
      const params = { page: 1, page_size: 200 };
      const result = validatePaginationParams(params);
      expect(result.page_size).toBe(100);
    });
  });

  describe('paginateArray', () => {
    const items = ['a', 'b', 'c', 'd', 'e', 'f', 'g'];

    it('should paginate first page', () => {
      const result = paginateArray(items, { page: 1, page_size: 3 });
      expect(result.items).toEqual(['a', 'b', 'c']);
      expect(result.current_page).toBe(1);
      expect(result.total_items).toBe(7);
      expect(result.total_pages).toBe(3);
      expect(result.has_next).toBe(true);
      expect(result.has_prev).toBe(false);
    });

    it('should paginate last page', () => {
      const result = paginateArray(items, { page: 3, page_size: 3 });
      expect(result.items).toEqual(['g']);
      expect(result.has_next).toBe(false);
      expect(result.has_prev).toBe(true);
    });

    it('should handle page beyond total', () => {
      const result = paginateArray(items, { page: 10, page_size: 3 });
      expect(result.items).toEqual([]);
      expect(result.current_page).toBe(10);
    });

    it('should handle empty array', () => {
      const result = paginateArray([], { page: 1, page_size: 3 });
      expect(result.items).toEqual([]);
      expect(result.total_items).toBe(0);
      expect(result.total_pages).toBe(1);
    });
  });

  describe('calculateOffset', () => {
    it('should calculate correct offsets', () => {
      expect(calculateOffset(1, 10)).toBe(0);
      expect(calculateOffset(2, 10)).toBe(10);
      expect(calculateOffset(3, 5)).toBe(10);
    });

    it('should handle invalid inputs', () => {
      expect(calculateOffset(-1, 10)).toBe(0);
      expect(calculateOffset(1, 0)).toBe(0);
    });
  });

  describe('calculateTotalPages', () => {
    it('should calculate total pages', () => {
      expect(calculateTotalPages(25, 10)).toBe(3);
      expect(calculateTotalPages(20, 10)).toBe(2);
      expect(calculateTotalPages(10, 10)).toBe(1);
    });

    it('should handle edge cases', () => {
      expect(calculateTotalPages(0, 10)).toBe(1);
      expect(calculateTotalPages(-5, 10)).toBe(1);
      expect(calculateTotalPages(25, 0)).toBe(1);
    });
  });

  describe('getPageInfo', () => {
    it('should return correct page info', () => {
      const info = getPageInfo(2, 10, 25);
      expect(info.start_item).toBe(11);
      expect(info.end_item).toBe(20);
      expect(info.has_next).toBe(true);
      expect(info.has_prev).toBe(true);
    });

    it('should handle last page', () => {
      const info = getPageInfo(3, 10, 25);
      expect(info.start_item).toBe(21);
      expect(info.end_item).toBe(25);
      expect(info.has_next).toBe(false);
    });

    it('should handle invalid inputs', () => {
      const info = getPageInfo(-1, 0, 25);
      expect(info.start_item).toBe(1);
      expect(info.end_item).toBe(10);
    });
  });

  describe('createPaginationMeta', () => {
    it('should create pagination metadata', () => {
      const meta = createPaginationMeta(2, 10, 25);
      expect(meta.current_page).toBe(2);
      expect(meta.page_size).toBe(10);
      expect(meta.total_items).toBe(25);
      expect(meta.total_pages).toBe(3);
      expect(meta.start_item).toBe(11);
      expect(meta.end_item).toBe(20);
    });
  });

  describe('cursor functions', () => {
    describe('encodeCursor', () => {
      it('should encode string cursor', () => {
        const encoded = encodeCursor('test123');
        expect(typeof encoded).toBe('string');
        expect(encoded.length).toBeGreaterThan(0);
      });

      it('should encode number cursor', () => {
        const encoded = encodeCursor(123);
        expect(typeof encoded).toBe('string');
      });
    });

    describe('decodeCursor', () => {
      it('should decode valid cursor', () => {
        const original = 'test123';
        const encoded = encodeCursor(original);
        const decoded = decodeCursor(encoded);
        expect(decoded).toBe(original);
      });

      it('should return null for invalid cursor', () => {
        const decoded = decodeCursor('invalid-base64!');
        expect(decoded).toBeNull();
      });

      it('should handle empty cursor', () => {
        const decoded = decodeCursor('');
        expect(decoded).toBeNull();
      });
    });
  });

  describe('page size functions', () => {
    describe('validatePageSize', () => {
      it('should validate page size within bounds', () => {
        expect(validatePageSize(20)).toBe(20);
        expect(validatePageSize(50, 10, 100)).toBe(50);
      });

      it('should apply minimum bound', () => {
        expect(validatePageSize(0)).toBe(1);
        expect(validatePageSize(5, 10, 100)).toBe(10);
      });

      it('should apply maximum bound', () => {
        expect(validatePageSize(200)).toBe(100);
        expect(validatePageSize(150, 10, 100)).toBe(100);
      });
    });

    describe('getDefaultPageSize', () => {
      it('should return default page size', () => {
        expect(getDefaultPageSize()).toBe(10);
      });
    });

    describe('getMaxPageSize', () => {
      it('should return max page size', () => {
        expect(getMaxPageSize()).toBe(100);
      });
    });
  });

  describe('validation functions', () => {
    describe('isValidPage', () => {
      it('should validate page numbers', () => {
        expect(isValidPage(1)).toBe(true);
        expect(isValidPage(5)).toBe(true);
        expect(isValidPage(0)).toBe(false);
        expect(isValidPage(-1)).toBe(false);
      });
    });

    describe('isValidPageSize', () => {
      it('should validate page sizes', () => {
        expect(isValidPageSize(10)).toBe(true);
        expect(isValidPageSize(50)).toBe(true);
        expect(isValidPageSize(100)).toBe(true);
        expect(isValidPageSize(0)).toBe(false);
        expect(isValidPageSize(101)).toBe(false);
      });
    });
  });

  describe('generatePageNumbers', () => {
    it('should generate all pages when total is small', () => {
      const pages = generatePageNumbers(1, 3, 5);
      expect(pages).toEqual([1, 2, 3]);
    });

    it('should generate pages with ellipsis for large ranges', () => {
      const pages = generatePageNumbers(5, 10, 5);
      expect(pages).toEqual([1, '...', 3, 4, 5, 6, 7, '...', 10]);
    });

    it('should handle edge cases', () => {
      expect(generatePageNumbers(1, 1, 5)).toEqual([1]);
      expect(generatePageNumbers(1, 0, 5)).toEqual([]);
    });
  });

  describe('getPageSizeOptions', () => {
    it('should return page size options', () => {
      const options = getPageSizeOptions();
      expect(options).toEqual([5, 10, 20, 50, 100]);
    });
  });

  describe('needsPagination', () => {
    it('should determine if pagination is needed', () => {
      expect(needsPagination(50, 10)).toBe(true);
      expect(needsPagination(10, 10)).toBe(false);
      expect(needsPagination(5, 10)).toBe(false);
    });
  });

  describe('getPaginationSummary', () => {
    it('should generate pagination summary', () => {
      const meta = {
        current_page: 2,
        page_size: 10,
        total_items: 25,
        total_pages: 3,
        has_next: true,
        has_prev: true,
        start_item: 11,
        end_item: 20,
      };
      const summary = getPaginationSummary(meta);
      expect(summary).toBe('Showing 11 to 20 of 25 items');
    });

    it('should handle no items', () => {
      const meta = {
        current_page: 1,
        page_size: 10,
        total_items: 0,
        total_pages: 1,
        has_next: false,
        has_prev: false,
        start_item: 0,
        end_item: 0,
      };
      const summary = getPaginationSummary(meta);
      expect(summary).toBe('No items');
    });
  });
});

File: aegis/backend/test/internal/utils/pagination_test.go
package utils_test

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/suite"

	"github.com/balkanid/aegis-backend/internal/utils"
)

type PaginationTestSuite struct {
	suite.Suite
}

func (suite *PaginationTestSuite) SetupTest() {
	// Setup code if needed
}

func (suite *PaginationTestSuite) TearDownTest() {
	// Cleanup code if needed
}

func TestPaginationTestSuite(t *testing.T) {
	suite.Run(t, new(PaginationTestSuite))
}

func (suite *PaginationTestSuite) TestDefaultPaginationParams() {
	params := utils.DefaultPaginationParams()

	assert.Equal(suite.T(), 1, params.Page)
	assert.Equal(suite.T(), 10, params.PageSize)
}

func (suite *PaginationTestSuite) TestValidatePaginationParams() {
	tests := []struct {
		name     string
		input    utils.PaginationParams
		expected utils.PaginationParams
	}{
		{
			name:     "valid params",
			input:    utils.PaginationParams{Page: 2, PageSize: 20},
			expected: utils.PaginationParams{Page: 2, PageSize: 20},
		},
		{
			name:     "negative page",
			input:    utils.PaginationParams{Page: -1, PageSize: 20},
			expected: utils.PaginationParams{Page: 1, PageSize: 20},
		},
		{
			name:     "zero page",
			input:    utils.PaginationParams{Page: 0, PageSize: 20},
			expected: utils.PaginationParams{Page: 1, PageSize: 20},
		},
		{
			name:     "negative page size",
			input:    utils.PaginationParams{Page: 1, PageSize: -5},
			expected: utils.PaginationParams{Page: 1, PageSize: 10},
		},
		{
			name:     "zero page size",
			input:    utils.PaginationParams{Page: 1, PageSize: 0},
			expected: utils.PaginationParams{Page: 1, PageSize: 10},
		},
		{
			name:     "page size too large",
			input:    utils.PaginationParams{Page: 1, PageSize: 200},
			expected: utils.PaginationParams{Page: 1, PageSize: 100},
		},
	}

	for _, tt := range tests {
		suite.T().Run(tt.name, func(t *testing.T) {
			result := utils.ValidatePaginationParams(tt.input)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func (suite *PaginationTestSuite) TestPaginateSlice() {
	items := []interface{}{"a", "b", "c", "d", "e", "f", "g", "h", "i", "j"}

	tests := []struct {
		name     string
		params   utils.PaginationParams
		expected utils.PaginationResult
	}{
		{
			name:   "first page",
			params: utils.PaginationParams{Page: 1, PageSize: 3},
			expected: utils.PaginationResult{
				Items:       []interface{}{"a", "b", "c"},
				CurrentPage: 1,
				PageSize:    3,
				TotalItems:  10,
				TotalPages:  4,
				HasNext:     true,
				HasPrev:     false,
			},
		},
		{
			name:   "middle page",
			params: utils.PaginationParams{Page: 2, PageSize: 3},
			expected: utils.PaginationResult{
				Items:       []interface{}{"d", "e", "f"},
				CurrentPage: 2,
				PageSize:    3,
				TotalItems:  10,
				TotalPages:  4,
				HasNext:     true,
				HasPrev:     true,
			},
		},
		{
			name:   "last page",
			params: utils.PaginationParams{Page: 4, PageSize: 3},
			expected: utils.PaginationResult{
				Items:       []interface{}{"j"},
				CurrentPage: 4,
				PageSize:    3,
				TotalItems:  10,
				TotalPages:  4,
				HasNext:     false,
				HasPrev:     true,
			},
		},
		{
			name:   "page beyond total",
			params: utils.PaginationParams{Page: 10, PageSize: 3},
			expected: utils.PaginationResult{
				Items:       []interface{}{},
				CurrentPage: 10,
				PageSize:    3,
				TotalItems:  10,
				TotalPages:  4,
				HasNext:     false,
				HasPrev:     true,
			},
		},
		{
			name:   "empty slice",
			params: utils.PaginationParams{Page: 1, PageSize: 3},
			expected: utils.PaginationResult{
				Items:       []interface{}{},
				CurrentPage: 1,
				PageSize:    3,
				TotalItems:  0,
				TotalPages:  1,
				HasNext:     false,
				HasPrev:     false,
			},
		},
	}

	for _, tt := range tests {
		suite.T().Run(tt.name, func(t *testing.T) {
			var input interface{}
			if tt.name == "empty slice" {
				input = []interface{}{}
			} else {
				input = items
			}
			result := utils.PaginateSlice(input, tt.params)
			assert.Equal(t, tt.expected.Items, result.Items)
			assert.Equal(t, tt.expected.CurrentPage, result.CurrentPage)
			assert.Equal(t, tt.expected.PageSize, result.PageSize)
			assert.Equal(t, tt.expected.TotalItems, result.TotalItems)
			assert.Equal(t, tt.expected.TotalPages, result.TotalPages)
			assert.Equal(t, tt.expected.HasNext, result.HasNext)
			assert.Equal(t, tt.expected.HasPrev, result.HasPrev)
		})
	}
}

func (suite *PaginationTestSuite) TestCalculateOffset() {
	tests := []struct {
		name     string
		page     int
		pageSize int
		expected int
	}{
		{"first page", 1, 10, 0},
		{"second page", 2, 10, 10},
		{"third page", 3, 5, 10},
		{"negative page", -1, 10, 0},
		{"zero page", 0, 10, 0},
		{"negative page size", 1, -5, 0},
		{"zero page size", 1, 0, 0},
	}

	for _, tt := range tests {
		suite.T().Run(tt.name, func(t *testing.T) {
			result := utils.CalculateOffset(tt.page, tt.pageSize)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func (suite *PaginationTestSuite) TestCalculateTotalPages() {
	tests := []struct {
		name       string
		totalItems int
		pageSize   int
		expected   int
	}{
		{"exact division", 20, 10, 2},
		{"remainder", 25, 10, 3},
		{"single page", 5, 10, 1},
		{"zero items", 0, 10, 1},
		{"negative items", -5, 10, 1},
		{"zero page size", 20, 0, 1},
		{"negative page size", 20, -5, 1},
	}

	for _, tt := range tests {
		suite.T().Run(tt.name, func(t *testing.T) {
			result := utils.CalculateTotalPages(tt.totalItems, tt.pageSize)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func (suite *PaginationTestSuite) TestGetPageInfo() {
	tests := []struct {
		name       string
		page       int
		pageSize   int
		totalItems int
		expected   utils.PageInfo
	}{
		{
			name:       "first page",
			page:       1,
			pageSize:   10,
			totalItems: 25,
			expected:   utils.PageInfo{StartItem: 1, EndItem: 10, HasNext: true, HasPrev: false},
		},
		{
			name:       "last page",
			page:       3,
			pageSize:   10,
			totalItems: 25,
			expected:   utils.PageInfo{StartItem: 21, EndItem: 25, HasNext: false, HasPrev: true},
		},
		{
			name:       "negative page",
			page:       -1,
			pageSize:   10,
			totalItems: 25,
			expected:   utils.PageInfo{StartItem: 1, EndItem: 10, HasNext: true, HasPrev: false},
		},
		{
			name:       "zero page size",
			page:       1,
			pageSize:   0,
			totalItems: 25,
			expected:   utils.PageInfo{StartItem: 1, EndItem: 10, HasNext: true, HasPrev: false},
		},
	}

	for _, tt := range tests {
		suite.T().Run(tt.name, func(t *testing.T) {
			result := utils.GetPageInfo(tt.page, tt.pageSize, tt.totalItems)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func (suite *PaginationTestSuite) TestCreatePaginationMeta() {
	tests := []struct {
		name       string
		page       int
		pageSize   int
		totalItems int
		expected   utils.PaginationMeta
	}{
		{
			name:       "normal case",
			page:       2,
			pageSize:   10,
			totalItems: 25,
			expected: utils.PaginationMeta{
				CurrentPage: 2,
				PageSize:    10,
				TotalItems:  25,
				TotalPages:  3,
				HasNext:     true,
				HasPrev:     true,
				StartItem:   11,
				EndItem:     20,
			},
		},
		{
			name:       "first page",
			page:       1,
			pageSize:   10,
			totalItems: 25,
			expected: utils.PaginationMeta{
				CurrentPage: 1,
				PageSize:    10,
				TotalItems:  25,
				TotalPages:  3,
				HasNext:     true,
				HasPrev:     false,
				StartItem:   1,
				EndItem:     10,
			},
		},
	}

	for _, tt := range tests {
		suite.T().Run(tt.name, func(t *testing.T) {
			result := utils.CreatePaginationMeta(tt.page, tt.pageSize, tt.totalItems)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func (suite *PaginationTestSuite) TestEncodeCursor() {
	tests := []struct {
		name     string
		id       uint
		expected string
	}{
		{"id 1", 1, "\x01"},
		{"id 42", 42, "*"},
		{"id 255", 255, "ÿ"},
	}

	for _, tt := range tests {
		suite.T().Run(tt.name, func(t *testing.T) {
			result := utils.EncodeCursor(tt.id)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func (suite *PaginationTestSuite) TestDecodeCursor() {
	tests := []struct {
		name     string
		cursor   string
		expected uint
		hasError bool
	}{
		{"valid cursor", "\x01", 1, false},
		{"empty cursor", "", 0, false},
		{"single char", "a", 97, false},
	}

	for _, tt := range tests {
		suite.T().Run(tt.name, func(t *testing.T) {
			result, err := utils.DecodeCursor(tt.cursor)
			if tt.hasError {
				assert.Error(t, err)
			} else {
				assert.NoError(t, err)
				assert.Equal(t, tt.expected, result)
			}
		})
	}
}

func (suite *PaginationTestSuite) TestValidatePageSize() {
	tests := []struct {
		name     string
		pageSize int
		minSize  int
		maxSize  int
		expected int
	}{
		{"within bounds", 20, 10, 50, 20},
		{"below min", 5, 10, 50, 10},
		{"above max", 60, 10, 50, 50},
		{"equal to min", 10, 10, 50, 10},
		{"equal to max", 50, 10, 50, 50},
	}

	for _, tt := range tests {
		suite.T().Run(tt.name, func(t *testing.T) {
			result := utils.ValidatePageSize(tt.pageSize, tt.minSize, tt.maxSize)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func (suite *PaginationTestSuite) TestGetDefaultPageSize() {
	result := utils.GetDefaultPageSize()
	assert.Equal(suite.T(), 10, result)
}

func (suite *PaginationTestSuite) TestGetMaxPageSize() {
	result := utils.GetMaxPageSize()
	assert.Equal(suite.T(), 100, result)
}

func (suite *PaginationTestSuite) TestIsValidPage() {
	tests := []struct {
		name     string
		page     int
		expected bool
	}{
		{"valid page", 1, true},
		{"valid page 2", 5, true},
		{"invalid negative", -1, false},
		{"invalid zero", 0, false},
	}

	for _, tt := range tests {
		suite.T().Run(tt.name, func(t *testing.T) {
			result := utils.IsValidPage(tt.page)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func (suite *PaginationTestSuite) TestIsValidPageSize() {
	tests := []struct {
		name     string
		pageSize int
		expected bool
	}{
		{"valid size", 10, true},
		{"valid size 2", 50, true},
		{"valid max", 100, true},
		{"invalid negative", -1, false},
		{"invalid zero", 0, false},
		{"invalid too large", 101, false},
	}

	for _, tt := range tests {
		suite.T().Run(tt.name, func(t *testing.T) {
			result := utils.IsValidPageSize(tt.pageSize)
			assert.Equal(t, tt.expected, result)
		})
	}
}